command.c

```diff
@@ -694,15 +694,37 @@ static const rarch_memory_descriptor_t* command_memory_get_descriptor(const rarc
       {
          /* if select is 0, attempt to explicitly match the address */
          if (address >= desc->core.start && address < desc->core.start + desc->core.len)
+         {
+            *offset = address - desc->core.start;
             return desc;
+         }
       }
       else
       {
          /* otherwise, attempt to match the address by matching the select bits */
          if (((desc->core.start ^ address) & desc->core.select) == 0)
          {
+            /* adjust the address to the start of the descriptor */
+            unsigned desc_offset = address - (unsigned)desc->core.start;
```

Comment:
I think there is an assumption here that the translation from emulated address to backing buffer address is "distributive". Currently, the main part of this translation, is the disconnect aka mmap_reduce function, so distributiveness would mean that:
```
mmap_reduce((address - start), disconnect) ==
mmap_reduce(address, disconnect) - mmap_reduce(start, disconnect)
```

However, this is not not mathematically true (even though equality does seem to be satisfied for the SNES example case). Counterexample:

```
          start    select   disconn      address
          81800    80800    0f0f0        90800
0x000 == mmap_reduce((address - start), disconnect)
 vs.
0x110 == mmap_reduce(address, disconnect) - mmap_reduce(start, disconnect)
```

Reply:
That seems logical, but this implementation matches the original implementation that you referenced: https://github.com/libretro/RetroArch/blob/v1.9.0/cheevos/fixup.c#L235-L239

I didn't include the additional masking as `disconnect_mask` was calculated from the `desc->core.len`, presumably to ensure the values didn't exceed the provided buffer. The direct check against `len` should satisfy that constraint, and was also present in the previous code after the mmap_reduce call.

Also, I find it hard to imagine that a starting address would have bits set that aren't part of the `select` mask.  If you change `start` to match `select` in your example, both values come out the same:
```
          start    select   disconn      address
          80800    80800    0f0f0        90800
0x100 == mmap_reduce((address - start), disconnect) // 90800-80800 = 10000 => reduce to 100
 vs.
0x100 == mmap_reduce(address, disconnect) - mmap_reduce(start, disconnect) // reduced: 980-880 = 100
```

The discrepancy does still exist if `select` and `disconnect` overlap:
```
          start    select   disconn      address
          81800    81800    0f0f0        90800
0x000 == mmap_reduce((address - start), disconnect) // 90800-81800 = 0F000 => reduce to 000
 vs.
0x100 == mmap_reduce(address, disconnect) - mmap_reduce(start, disconnect) // reduced: 980-880 = 100
```

But even with an awkward `start`, a more logical `disconnect` works:
``` 
          start    select   disconn      address
          81800    81800    0e0f0        90800
0x100 == mmap_reduce((address - start), disconnect) // 90800-81800 = F000 => reduce to 100
 vs.
0x100 == mmap_reduce(address, disconnect) - mmap_reduce(start, disconnect) // reduced: 1280-1180 = 100

          1001 0000 1000 0000 0000 address
          0000 1110 0000 1111 0000 disconn
          1001    0 1000      0000 => 1 0010 1000 0000 => 1280

          1000 0001 1000 0000 0000 start
          0000 1110 0000 1111 0000 disconn
          1000    1 1000      0000 => 1 0001 1000 0000 => 1180
```

Your proposal does seem like it would handle more abstract cases, but this matches the old behavior, and has already been merged. Do you think this is worth opening another PR?

Reply:
Good point, the implementation does match the behavior from 1.9.0, which may not have had any issues reported. (I only noticed this math oddity when reading the new code - probably because the new variable names are clearer!)

I also wasn't able to invent a plausible example that would exhibit this issue in a reasonable amount of time. Thank you for looking for one as well.

I'm OK with the fix as-is. By the way, my testing looks good too.